### ISSUE TITLE
Refactor protobufs

### ISSUE THREAD

**Issue by ursoft** [zoffline/zwift-offline#157]
Just decompiled all used protobufs from the 100317+ binary code and put results into my fork https://github.com/ursoft/zwift-offline
Original names (from debug info and strings) for messages, duplicates and other things are commented inside corresponding proto-files. Some data types may be decompiled with errors (intXX<->sintXX<->uintXX, string<->bytes etc), so all changes need to be checked on real data.

**Comment by oldnapalm** [comment]
I've been using the refactored protobufs, no issues so far. Also have various success reports of people using the zca branch.

Is there any specific test you recommend or just normal use?

Thanks for the awesome work.

**Comment by ursoft** [comment]
> I only have one trainer so it's difficult to test riding with a friend.

You need Android app named "nRF Connect". Create GATT Server for Cycling Power in nRF Connect or use my cfg:
```
<server-configuration name="Power80W">
   <service uuid="00001818-0000-1000-8000-00805f9b34fb">
      <characteristic uuid="00002a63-0000-1000-8000-00805f9b34fb" value="00005000">
         <descriptor configure="CCCD"/>
         <permission name="READ"/>
         <property name="NOTIFY"/>
      </characteristic>
   </service>
</server-configuration>
```
also you need a macro for nRF Connect:
```
<macro name="Send Power notifications every second" icon="UPLOAD">
   <sleep description="This sample will show how to automate sending every second a notification from the Android device." timeout="300" />
   <sleep description="Make sure the connected device enabled CCCD for 00002a63-0000-1000-8000-00805f9b34fb characteristic." timeout="200" />
   <send-notification description="Send 0x00005000 as notification from 00002a63-0000-1000-8000-00805f9b34fb." characteristic-uuid="00002a63-0000-1000-8000-00805f9b34fb" service-uuid="00001818-0000-1000-8000-00805f9b34fb" value="00005000"/>
   <sleep description="Now sleep some time, so the notification is sent every second." timeout="400"/>
   <sleep description="To run the macro in a loop check the &#8644; option above, and run macro with &#9654; button." timeout="100"/>
</macro>
```

**Comment by ursoft** [comment]
> noticed you changed PlayerState.heading from int64 to uint32

Thanks, it's my bug, need to be fixed. In binary code:
```
      case 13:
          ...
          done = (google::protobuf::internal *)google::protobuf::internal::VarintParseSlow64(
                                                 (google::protobuf::internal *)TagFallback,
                                                 (const char *)a2,
                                                 (unsigned int)a3);
          this->m_heading = (int)a2; // 64 -> 32 bit
```
this code was autogenerated from proto and I do not know why this->m_heading is truncated after read...

**Comment by oldnapalm** [comment]
Merged your latest changes. Just missed one change here after route changed to 32 bit (otherwise Watopia pace partners will show with null route) https://github.com/ursoft/zwift-offline/commit/8356bef1b6fc059b626ecb5406e5ea3323023c4c#diff-f4053a7fcb3c73ca8b58c6fb646b8f0e2c467edfad4759aa3a241f09b3fe2ca2L315

**Comment by ursoft** [comment]
Let's call that bug "Peer disappear" (tried to reproduce it today with no luck), I'll try to describe it here: https://github.com/ursoft/zwift-offline/issues/1

Another "interesting" issue about protobufs or profile I call "Circus" https://github.com/ursoft/zwift-offline/issues/2

**Comment by ursoft** [comment]
> I mean, after Attribute.id changed to 64 bit
As I can see in zwift::protobuf::ProfileAttribute::_InternalParse, ProfileAttribute.Id  is 32 bit
But in previous named PacerSetting it was 64 bit too...

**Comment by oldnapalm** [comment]
> As I can see in zwift::protobuf::ProfileAttribute::_InternalParse, ProfileAttribute.Id is 32 bit But in previously named 'PacerSetting' it was 64 bit too...

Yes, but with 'PacerSetting' we checked `if id == 3273955058` for Watopia pace partners and when you changed to 32 bit you also changed to `if id == -1021012238`

> Another "interesting" issue about protobufs or profile I call "Circus" [ursoft#2](https://github.com/ursoft/zwift-offline/issues/2)

If it's not related to steering, maybe it's about the packets synchronization. If you receive a PlayerState packet with given speed and roadTime values for a remote player, and the next packet has a roadTime value which is not consistent with the "predicted" position based on the previous packet, the avatar will make these maneuvers to try to synchronize the position. That's what I tried to fix with the change of updating the online players states only once a second, but maybe it wasn't a good approach, maybe online players should be updated more often.
